papers/scipy/scipy.rst

```diff
@@ -0,0 +1,1706 @@
+    colors of the filtered signal indicate batches that
+    were processed in separate calls to ``sosfilt``.
+    :label:`fig-bandpass-batch-example`
+
+Solving linear recurrence relations
+-----------------------------------
+
+Variations of the question::
+
+        How do I speed up the following calculation?
+
+        y[i+1] = alpha*y[i] + c*x[i]
+
+often arise on mailing lists and online forums.  Sometimes more
+terms such as ``beta*y[i-1]`` or ``d*x[i-1]`` are included on the right.
+These recurrence relations show up in, for example, GARCH models
+and other linear stochastic models.
+Such a calculation can be written in the form of Eqn.
+(:ref:`eq-filter-recurrence`), so a solution can be computed
+using ``lfilter``.
+
+Here's an example that is similar to several questions that
+have appeared on the programming Q&A website ``stackoverflow.com``.
+The one-dimensional array  ``h`` is an input, and ``alpha``, ``beta`` and
+``gamma`` are constants::
+
+    y = np.empty(len(h))
+    y[0] = alpha
+    for i in np.arange(1, len(h)):
+        y[i] = alpha + beta*y[i-1] + gamma*h[i-1]
+
+To use ``lfilter`` to solve the problem, we have to translate
+the linear recurrence::
+
+    y[i] = alpha + beta*y[i-1] + gamma*h[i-1]
+
+into the form of Eqn. (:ref:`eq-filter-recurrence`), which will give us the
+coefficients ``b`` and ``a`` of the transfer function.  Define::
+
+    x[i] = alpha + gamma*h[i]
+
+so the recurrence relation is::
+
+    y[i] = x[i-1] + beta*y[i-1]
+
+Compare this to Eqn. (:ref:`eq-filter-recurrence`); 
+we see that :math:`a_{_0} = 1`, :math:`a_{_1} = -\rm{beta}`,
+:math:`b_{_0} = 0` and :math:`b_{_1} = 1`.
+So we have our transfer function coefficients::
+
+    b = [0, 1]
+    a = [1, -beta]
+
+We also have to ensure that the initial condition is set correctly to
+reproduce the desired calculation.
+We want the initial condition to be set as if we had values ``x[-1]``
+and ``y[-1]``, and ``y[0]`` is computed using the recurrence relation.
+Given the above recurrence relation, the formula for ``y[0]`` is::
+
+    y[0] = x[-1] + beta*y[-1]
+
+We want ``y[0]`` to be ``alpha``, so we'll set ``y[-1] = 0`` and
+``x[-1] = alpha``.  To create initial conditions for ``lfilter``
+that will set up the filter to act like it had just operated on
+those previous values, we use ``scipy.signal.lfiltic``::
+
+    zi = lfiltic(b, a, y=[0], x=[alpha])
+
+The ``y`` and ``x`` arguments are the "previous" values that will
+be used to set the initial conditions.  In general, one sets
+``y=[y[-1], y[-2], ..]`` and ``x=[x[-1], x[-2], ...]``, giving as
+many values as needed to determine the initial condition for
+``lfilter``.  In this example, we have just one previous value
+for ``y`` and ``x``.
+
+Putting it all together, here is the code using ``lfilter`` that
+replaces the for-loop shown above::
+
+    b = [0, 1]
+    a = [1, -beta]
+    zi = lfiltic(b, a, y=[0], x=[alpha])
+    y, zo = lfilter(b, a, alpha + gamma*h, zi=zi)
+
```

Comment:
Suggest to conclude with ``Result: X times faster for a signal of size N.``

Reply:
Not really needed--the motivation isn't speed.